File: software_challenge_sdk/src/server_api/sc/framework/plugins/RoundBasedGameInstance.java
package sc.framework.plugins;

import java.util.*;
import java.util.Map.Entry;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.thoughtworks.xstream.annotations.XStreamImplicit;
import com.thoughtworks.xstream.annotations.XStreamOmitField;

import sc.api.plugins.IGameInstance;
import sc.api.plugins.exceptions.GameLogicException;
import sc.api.plugins.host.IGameListener;
import sc.protocol.responses.ProtocolErrorMessage;
import sc.protocol.responses.ProtocolMessage;
import sc.protocol.responses.ProtocolMove;
import sc.shared.InvalidMoveException;
import sc.shared.PlayerScore;
import sc.shared.ScoreCause;
import sc.shared.WinCondition;

public abstract class RoundBasedGameInstance<P extends SimplePlayer> implements IGameInstance {
  private static Logger logger = LoggerFactory
          .getLogger(RoundBasedGameInstance.class);
  protected P activePlayer = null;

  /*
   * round equals the turn in the GameState, not the round the game is currently in
   * TODO rename round to turn
   */
  private int round = 0;

  @XStreamOmitField
  private Optional<Integer> paused = Optional.empty();

  @XStreamOmitField
  private Runnable afterPauseAction = null;

  @XStreamOmitField
  private Object afterPauseLock = new Object();

  @XStreamOmitField
  private ActionTimeout requestTimeout = null;

  @XStreamOmitField
  protected final List<IGameListener> listeners = new LinkedList<>();

  @XStreamImplicit(itemFieldName = "player")
  protected final List<P> players = new ArrayList<>();

  @XStreamOmitField
  protected String pluginUUID;

  public int getRound() {
    return this.round;
  }

  /**
   * Called by the Server once an action was received.
   *
   * @param fromPlayer The player who invoked this action.
   * @param data       The plugin-specific data.
   *
   * @throws GameLogicException if any invalid action is done, i.e. game rule violation
   */
  public final void onAction(SimplePlayer fromPlayer, ProtocolMessage data)
          throws GameLogicException {
    Optional<String> errorMsg = Optional.empty();
    if (fromPlayer.equals(this.activePlayer)) {
      if (wasMoveRequested()) {
        this.requestTimeout.stop();

        if (this.requestTimeout.didTimeout()) {
          logger.warn("Client hit soft-timeout.");
          fromPlayer.setSoftTimeout(true);
          onPlayerLeft(fromPlayer, ScoreCause.SOFT_TIMEOUT);
        } else {
          if (!isPaused()){
            onRoundBasedAction(fromPlayer, data);
          } else {
            logger.info("Game is paused. Save this action.");
            final P currentActivePlayer = this.activePlayer;
            synchronized (this.afterPauseLock) {
              logger.debug("Setting AfterPauseAction");

              this.afterPauseAction = () -> {
                requestMove(currentActivePlayer);
                next();
              };

              for (IGameListener listener : this.listeners) {
                listener.onPaused(currentActivePlayer);
              }
            }
          }
        }
      } else {
        errorMsg = Optional.of("We didn't request a data from you yet.");
      }
    } else {
      errorMsg = Optional.of("It's not your turn yet.");
    }
    if (errorMsg.isPresent()) {
      fromPlayer.notifyListeners(new ProtocolErrorMessage(data, errorMsg.get()));
      throw new GameLogicException(errorMsg.get());
    }
  }

  private boolean wasMoveRequested() {
    return this.requestTimeout != null;
  }

  protected abstract void onRoundBasedAction(SimplePlayer fromPlayer, ProtocolMessage data)
          throws GameLogicException;

  /**
   * Checks if a win condition in the current game state is met.
   * Checks round limit and end of round (and playerStats).
   * Checks if goal is reached
   *
   * @return WinCondition with winner and reason or null, if no win condition is
   *         yet met.
   */
  protected abstract WinCondition checkWinCondition();

  /**
   * At any time this method might be invoked by the server. Any open handles
   * should be removed. No events should be sent out (GameOver etc) after this
   * method has been called.
   */
  public void destroy() {
    logger.info("Destroying Game");

    if (this.requestTimeout != null) {
      this.requestTimeout.stop();
      this.requestTimeout = null;
    }
  }

  /**
   * Server or an administrator requests the game to start now.
   */
  public void start() {
    next(this.players.get(0));
  }

  /**
   * On violation player is removed forcefully, if player has not violated, he has left by himself (i.e. Exception)
   *
   * @param player left player
   */
  public void onPlayerLeft(SimplePlayer player) {
    if (!player.hasViolated()) {
      player.setLeft(true);
      onPlayerLeft(player, ScoreCause.LEFT);
    } else {
      onPlayerLeft(player, ScoreCause.RULE_VIOLATION);
    }
  }

  /**
   * Handle leave of player
   */
  public void onPlayerLeft(SimplePlayer player, ScoreCause cause) {
    Map<SimplePlayer, PlayerScore> res = generateScoreMap();

    for (Entry<SimplePlayer, PlayerScore> entry : res.entrySet()) {
      PlayerScore score = entry.getValue();

      if (entry.getKey() == player) {
        score.setCause(cause);
      }
    }

    notifyOnGameOver(res);
  }

  protected void onActivePlayerChanged(P newActivePlayer) {
    // optional callback
  }

  protected void next() {
    next(getPlayerAfter(this.activePlayer));
  }

  protected final P getPlayerAfter(P player) {
    return getPlayerAfter(player, 1);
  }

  protected final P getPlayerAfter(P player, int step) {
    int playerPos = this.players.indexOf(player);
    playerPos = (playerPos + step) % this.players.size();
    return this.players.get(playerPos);
  }

  protected final void next(P nextPlayer) {
    if (increaseTurnIfNecessary(nextPlayer)) {
      this.round++;

      // change player before calling new round callback
      this.activePlayer = nextPlayer;
    }
    logger.debug("next round ({}) for player {}", this.round, nextPlayer);

    this.activePlayer = nextPlayer;
    notifyOnNewState(getCurrentState());

    if (checkWinCondition() != null) {
      notifyOnGameOver(generateScoreMap());
    } else {
      if (!this.isPaused()) {
        notifyActivePlayer();
      }
    }
  }

  public abstract PlayerScore getScoreFor(P p);

  protected boolean increaseTurnIfNecessary(P nextPlayer) {
    return (this.activePlayer != nextPlayer && this.players
            .indexOf(nextPlayer) == 0);
  }

  /**
   * Gets the current state representation.
   *
   * @return current state
   */
  protected abstract Object getCurrentState();

  /**
   * Notifies the active player that it's his/her time to make a move.
   */
  protected final void notifyActivePlayer() {
    requestMove(activePlayer);
  }

  /**
   * Sends a MoveRequest directly to the player (does not take PAUSE into
   * account)
   *
   * @param player player to make a move
   */
  protected synchronized final void requestMove(P player) {
    final ActionTimeout timeout = player.isCanTimeout() ? getTimeoutFor(player)
            : new ActionTimeout(false);

    final Logger logger = RoundBasedGameInstance.logger;
    final P playerToTimeout = player;

    // Signal the JVM to do a GC run now and lower the propability that the GC
    // runs when the player sends back its move, resulting in disqualification
    // because of soft timeout.
    System.gc();

    this.requestTimeout = timeout;
    timeout.start(() -> {
      logger.warn("Player {} reached the timeout of {}ms",
              playerToTimeout, timeout.getHardTimeout());
      playerToTimeout.setHardTimeout(true);
      onPlayerLeft(playerToTimeout, ScoreCause.HARD_TIMEOUT);
    });

    player.requestMove();
  }

  protected ActionTimeout getTimeoutFor(P player) {
    return new ActionTimeout(true);
  }

  protected final boolean isPaused() {

    return this.paused.isPresent() || this.paused.get() > this.round;
  }


  public void afterPause() {
    synchronized (this.afterPauseLock) {
      if (this.afterPauseAction == null) {
        logger
                .error("AfterPauseAction was null. Might cause a deadlock.");
      } else {
        logger.info("Run AfterPauseAction.");
        Runnable action = this.afterPauseAction;
        this.afterPauseAction = null;
        action.run();
      }
    }
  }

  /**
   * Pauses game
   *
   * @param pause true if game should be paused
   */
  public void setPauseMode(Optional<Integer> pause) {
    this.paused = pause;
  }

  public Map<SimplePlayer, PlayerScore> generateScoreMap() {
    Map<SimplePlayer, PlayerScore> map = new HashMap<SimplePlayer, PlayerScore>();

    for (final P p : this.players) {
      map.put(p, getScoreFor(p));
    }

    return map;
  }

  // XXX methods from former SimpleGameInstance

  /**
   * Extends the set of listeners.
   *
   * @param listener GameListener to be added
   */
  public void addGameListener(IGameListener listener) {
    this.listeners.add(listener);
  }

  /**
   * Removes listener XXX is this right/complete?
   *
   * @param listener GameListener to be removed
   */
  public void removeGameListener(IGameListener listener) {
    this.listeners.remove(listener);
  }

  protected void notifyOnGameOver(Map<SimplePlayer, PlayerScore> map) {
    for (IGameListener listener : this.listeners) {
      try {
        listener.onGameOver(map);
      } catch (Exception e) {
        logger.error("GameOver Notification caused an exception.", e);
      }
    }
  }

  protected void notifyOnNewState(Object mementoState) {
    for (IGameListener listener : this.listeners) {
      logger.debug("notifying {} about new game state", listener);
      try {
        listener.onStateChanged(mementoState);
      } catch (Exception e) {
        logger.error("NewState Notification caused an exception.", e);
      }
    }
  }

  /**
   * Catch block, after an invalid move was performed
   * @param e catched Exception
   * @param author player, that caused the exception
   * @throws GameLogicException Always thrown
   */
  public void catchInvalidMove(InvalidMoveException e, SimplePlayer author) throws GameLogicException {
    author.setViolated(true);
    String err = "Ungueltiger Zug von '" + author.getDisplayName() + "'.\n" + e.getMessage();
    author.setViolationReason(e.getMessage());
    logger.error(err, e);
    author.notifyListeners(new ProtocolErrorMessage(e.getMove(), err));
    throw new GameLogicException(err);
  }

  public String getPluginUUID() {
    return pluginUUID;
  }
}

File: socha-sdk/src/server-api/sc/framework/plugins/RoundBasedGameInstance.kt
package sc.framework.plugins;

import com.thoughtworks.xstream.annotations.XStreamImplicit
import com.thoughtworks.xstream.annotations.XStreamOmitField
import org.slf4j.LoggerFactory
import sc.api.plugins.IGameInstance
import sc.api.plugins.IGameState
import sc.api.plugins.exceptions.GameLogicException
import sc.api.plugins.host.IGameListener
import sc.protocol.responses.ProtocolErrorMessage
import sc.protocol.responses.ProtocolMessage
import sc.shared.InvalidMoveException
import sc.shared.PlayerScore
import sc.shared.ScoreCause
import sc.shared.WinCondition
import java.lang.NullPointerException
import java.util.*

abstract class RoundBasedGameInstance : IGameInstance {
    companion object {
        val logger = LoggerFactory.getLogger(RoundBasedGameInstance::class.java)
    }
    
    protected var activePlayer: Player? = null
    
    private var turn: Int = 0
    
    @XStreamOmitField
    private var paused = Optional.empty<Int>()
    
    @XStreamOmitField
    private var requestTimeout: ActionTimeout? = null
    
    @XStreamOmitField
    protected val listeners = mutableListOf<IGameListener>()
    
    @XStreamImplicit(itemFieldName = "player")
    override val players = mutableListOf<Player>()
    
    fun getRound(): Int = turn / 2
    
    /**
     * Called by the Server once an action was received.
     *
     * @param fromPlayer The player who invoked this action.
     * @param data       The plugin-specific data.
     *
     * @throws GameLogicException if any invalid action is done, i.e. game rule violation
     */
    override fun onAction(fromPlayer: Player, data: ProtocolMessage) {
        var errorMsg: String? = null
        if (fromPlayer == activePlayer) {
            if (requestTimeout != null) {
                requestTimeout!!.stop()
                
                if (requestTimeout!!.didTimeout()) {
                    logger.warn("Client hit soft-timeout.")
                    fromPlayer.softTimeout = true
                    onPlayerLeft(fromPlayer, ScoreCause.SOFT_TIMEOUT)
                } else {
                    onRoundBasedAction(fromPlayer, data)
                }
            } else {
                errorMsg = "We didn't request a move from you yet"
            }
        } else {
            errorMsg = "It's not your turn yet."
        }
        if (errorMsg != null) {
            fromPlayer.notifyListeners(ProtocolErrorMessage(data, errorMsg))
            throw GameLogicException(errorMsg)
        }
    }
    
    protected abstract fun onRoundBasedAction(fromPlayer: Player, data: ProtocolMessage)
    
    protected abstract fun checkWinCondition(): WinCondition?
    
    override fun destroy() {
        logger.info("Destroying Game")
        requestTimeout?.stop()
        requestTimeout = null
    }
    
    override fun start() {
        next(players.first(), true)
    }
    
    override fun onPlayerLeft(player: Player, cause: ScoreCause?) {
        if (cause == ScoreCause.REGULAR)
            return
        
        cause.let {
            when (player.violated) {
                true -> ScoreCause.RULE_VIOLATION
                false -> {
                    player.left = true
                    ScoreCause.LEFT
                }
            }
        }
        val scores = generateScoreMap()
        scores.getValue(player).cause = cause
        notifyOnGameOver(scores)
    }
    
    fun next(nextPlayer: Player, firstTurn: Boolean = false) {
        logger.debug("next round (${getRound()}) for player $nextPlayer")
        if (!firstTurn)
            turn++
        activePlayer = nextPlayer
        // don't notify on new state if game is paused or client may begin to calculate something
        if (!isPaused())
            notifyOnNewState(getCurrentState())
    }
    
    abstract fun getScoreFor(player: Player): PlayerScore
    
    protected abstract fun getCurrentState(): IGameState
    
    protected fun notifyActivePlayer() {
        try {
            requestMove(activePlayer!!)
        } catch (e: NullPointerException) {
            logger.error("Can't notify active player because there is no player")
            throw e
        }
    }
    
    protected fun requestMove(player: Player) {
        val timeout =
                if (player.isCanTimeout) getTimeoutFor(player)
                else ActionTimeout(false)
        
        val logger = logger
        val playerToTimeout = player
        
        // Signal the JVM to do a GC run now and lower the probability that the GC
        // runs when the players sends back its move, resulting in disqualification
        // because of soft timeout
        System.gc()
        
        requestTimeout = timeout
        timeout.start {
            logger.warn("Player $playerToTimeout reached the timeout of ${timeout.hardTimeout}ms")
            playerToTimeout.hardTimeout = true
            onPlayerLeft(playerToTimeout, ScoreCause.HARD_TIMEOUT)
        }
        player.requestMove()
    }
    
    protected open fun getTimeoutFor(player: Player) = ActionTimeout(true)
    
    fun isPaused():Boolean =
            paused.map { it <= turn }.orElse(false)
    
    fun afterPause() {
        logger.info("Sending MoveRequest to player $activePlayer")
        notifyOnNewState(getCurrentState())
        notifyActivePlayer()
    }
    
    fun setPauseMode(pause: Boolean) {
        paused = if (pause) {
            if (requestTimeout != null) Optional.of(turn + 1)
            else Optional.of(turn)
        } else Optional.empty()
    }
    
    fun generateScoreMap(): Map<Player, PlayerScore> =
            players.map {
                it to getScoreFor(it)
            }.toMap()
    
    override fun addGameListener(listener: IGameListener) {
        listeners.add(listener)
    }
    
    override fun removeGameListener(listener: IGameListener) {
        listeners.remove(listener)
    }
    
    protected fun notifyOnGameOver(map: Map<Player, PlayerScore>) {
        for (listener in listeners) {
            try {
                listener.onGameOver(map)
            } catch (e: Exception) {
                logger.error("GameOver Notification caused an exception.", e)
            }
        }
    }
    
    protected fun notifyOnNewState(mementoState: IGameState) {
        for (listener in listeners) {
            logger.debug("notifying $listener about new game state")
            try {
                listener.onStateChanged(mementoState)
            } catch (e: Exception) {
                logger.error("NewState Notification caused an exception.", e)
            }
        }
    }
    
    fun catchInvalidMove(e: InvalidMoveException, author: Player) {
        author.violated = true
        val errorMsg = "Ungueltiger Zug von ${author.displayName}'.\n$e"
        author.violationReason = e.message
        logger.error(errorMsg)
        author.notifyListeners(ProtocolErrorMessage(e.move, errorMsg))
        throw e
    }
    
    fun addPlayer(player: Player) {
        players.add(player)
    }
}
